RoutePattern everywhere

rynowak opened the pull request.

rynowak commented on samples/DispatcherSample.Web/Startup.cs:
```diff
@@ -39,10 +40,8 @@ public void ConfigureServices(IServiceCollection services)
+                        RoutePatternFactory.Parse("/"),
                         new RouteValueDictionary(),
-                        new RouteValueDictionary(),
-                        new List<MatchProcessorReference>(),
```
I'm open to discussion about what constructors we should have here. I'm also sensitive to the fact that we have like 9 parameters. We should try to make a decision here that will carry us through preview1.

JamesNK replied:
Will your average user ever want to create an endpoint themselves? If so then maybe one or two shorter ones? I don't have any strong opinions.

rynowak replied:
I don't think that's likely in 2.2 the audience for now is frameworks/middlewarelikes.

JamesNK commented on src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs:
```diff
@@ -20,145 +20,120 @@ internal class DefaultMatchProcessorFactory : MatchProcessorFactory
+            (var name, var argument) = Parse(value);
 
-            if (!_options.ConstraintMap.TryGetValue(constraintName, out var constraintType))
+            if (!_options.ConstraintMap.TryGetValue(name, out var type))
```
type -> constraintType

rynowak replied:
It's pretty clear right? What other type would it get confused with?

JamesNK replied:
sure. probably should have `nit:`ed it 😄

JamesNK commented on src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs:
```diff
@@ -20,145 +20,120 @@ internal class DefaultMatchProcessorFactory : MatchProcessorFactory
+                typeof(IRouteConstraint),
+                typeof(MatchProcessor));
+            throw new InvalidOperationException(message);
+
```
nit: remove

JamesNK commented on src/Microsoft.AspNetCore.Routing/Matchers/RouteConstraintMatchProcessor.cs:
```diff
@@ -0,0 +1,63 @@
+        public string ParameterName { get; }
+
+        public IRouteConstraint Constraint { get; }
+
```
nit: too liney

rynowak replied:
Fixed but not showing in the diff

JamesNK commented on src/Microsoft.AspNetCore.Routing/Matchers/RouteConstraintMatchProcessor.cs:
```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+
```
nit: too liney

JamesNK commented on src/Microsoft.AspNetCore.Routing/Matchers/RouteConstraintMatchProcessor.cs:
```diff
@@ -0,0 +1,63 @@
+
+namespace Microsoft.AspNetCore.Routing.Matchers
+{
+    internal class RouteConstraintMatchProcessor: MatchProcessor
```
nit: space

JamesNK commented on src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs:
```diff
@@ -92,6 +129,37 @@ private static int ComputeOutboundPrecedenceDigit(TemplateSegment segment)
+            var part = pathSegment.Parts[0];
+            if (part.IsLiteral)
+            {
+                return 5;
```
why 5 (and why 4 above it)

rynowak replied:
See comments elsewhere in this file.

rynowak replied:
Sorry, I should have explained this. None of what's here is new code. It's existing logic that can't be super-re-used without creating a nother overload.

JamesNK commented on src/Microsoft.AspNetCore.Routing/Template/RoutePrecedence.cs:
```diff
@@ -92,6 +129,37 @@ private static int ComputeOutboundPrecedenceDigit(TemplateSegment segment)
+            else if (part is RoutePatternParameterPart parameterPart)
+            {
+                Debug.Assert(parameterPart != null);
+                var digit = parameterPart.IsCatchAll ? 1 : 3;
```
why these numbers too

rynowak replied:
See comments elsewhere in this file

JamesNK commented on test/Microsoft.AspNetCore.Routing.Tests/Matchers/RouteConstraintMatchProcessorTest.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace Microsoft.AspNetCore.Routing.Matchers
+{
+    public class RouteConstraintMatchProcessorTest
```
hmm?

rynowak replied:
Oh oops

rynowak commented on test/Microsoft.AspNetCore.Routing.Tests/DefaultLinkGeneratorTest.cs:
```diff
@@ -785,50 +748,6 @@ public void GetLink_OptionalParameter_ParameterNotPresentInValues()
-        }
-
-        [Fact]
-        public void GetLink_OptionalParameter_ParameterNotPresentInValues_PresentInDefaults()
```
These two tests are not valid, and the fact that there were possible to write is a bug that I've fixed in RoutePattern. You can't have a default value for an optional parameter.

rynowak commented:
Updated. I added a few tersts for the RCMP. They are the gross Moq style, and I hate that but I wanted there to be **something** that unit tests that class.

kichalla commented on src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs:
```diff
@@ -20,145 +20,119 @@ internal class DefaultMatchProcessorFactory : MatchProcessorFactory
         {
-            var routeConstraint = DefaultInlineConstraintResolver.CreateConstraint(constraintType, constraintArgument);
-            var matchProcessor = new MatchProcessorReference(parameterName, routeConstraint).MatchProcessor;
+            var matchProcessor = (MatchProcessor)new RouteConstraintMatchProcessor(parameterName, constraint);
```
After this, call into the other InitializeMatchProcessor method

kichalla commented on src/Microsoft.AspNetCore.Routing/Matchers/DefaultMatchProcessorFactory.cs:
```diff
@@ -20,145 +20,119 @@ internal class DefaultMatchProcessorFactory : MatchProcessorFactory
             }
 
-            if (matchProcessorReference.MatchProcessor != null)
+            return InitializeMatchProcessor(parameterName, optional, value, null);
```
nit: named argument
